packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -1050,21 +1063,130 @@ class RenderParagraph extends RenderBox
             final TextRange originalRange = infoAttribute.range;
             attributes.add(
               infoAttribute.copy(
-                  range: TextRange(start: offset + originalRange.start,
-                      end: offset + originalRange.end)
+                range: TextRange(
+                  start: offset + originalRange.start,
+                  end: offset + originalRange.end,
+                ),
               ),
             );
           }
           buffer.write(label);
           offset += label.length;
         }
-        _cachedAttributedLabel = AttributedString(buffer.toString(), attributes: attributes);
+        _cachedAttributedLabels = <AttributedString>[AttributedString(buffer.toString(), attributes: attributes)];
       }
-      config.attributedLabel = _cachedAttributedLabel!;
+      config.attributedLabel = _cachedAttributedLabels![0];
       config.textDirection = textDirection;
     }
   }
 
+  ChildSemanticsConfigurationsResult _childSemanticsConfigurationsDelegate(List<SemanticsConfiguration> childConfigs) {
+    if (_cachedAttributedLabels == null) {
+      return _buildChildSemanticsConfigurationsResultAndFillCache(childConfigs);
+    }
+    return _buildChildSemanticsConfigurationsResultFromCache(childConfigs);
+  }
+
+  ChildSemanticsConfigurationsResult _buildChildSemanticsConfigurationsResultAndFillCache(List<SemanticsConfiguration> childConfigs) {
```

Comment:
How are _buildChildSemanticsConfigurationsResultAndFillCache and _buildChildSemanticsConfigurationsResultAndFillCache kept in sync to make sure they end up producing the same ChildSemanticsConfigurationsResult?

Reply:
added assert

Reply:
This doesn't really address the maintainability problem. If we ever have to fix anything in this logic, one would have to do it twice. Plus, the (complicated) assert doesn't really inspire confidence that the two are actually equal. For example, in some cases it just assumes configs are equal when the labels match.

Reply:
ah yes, I tried to combine this two functions in the previous commit, but the code looks like a trainwreck.

so I guess what you meant earlier is to make it more readable than to separate them out cache version and non-cache version?

Reply:
Yes, sorry if I wasn't clear...